proposals/0000-wire-protocol.md

```diff
@@ -0,0 +1,485 @@
+    optional uint64 nodes = 4;
+}
+```
+
+#### Cancel
+[msg-cancel]: #msg-cancel
+
+`type=8` Cancel a Request message sent earlier. The remote should react by aborting any active or queued Request message which matches the `index`, `bytes`, and `hash` parameters.
+
+```
+message Cancel {
+    required uint64 index = 1;
+    optional uint64 bytes = 2;
+    optional bool hash = 3;
+}
+```
+
+#### Data
+[msg-data]: #msg-data
+
+`type=9` Send a Hypercore feed block. May contain the data of the block, and may contain the hashes of the ancestor nodes in the merkle tree. Should only be sent in reaction to a Request message.
+
+When a Data message is received, its node hashes and signature should be verified against any local tree information. If the nodes can be verified, they and the block data may be stored locally.
+
+If a Data message is received for a block which was not requested, the peer can react by ignoring the data and sending an Unhave message in response. This will inform the remote that the data was rejected, and is not stored.
+
+```
+message Data {
+    message Node {
+        required uint64 index = 1;
+        required bytes hash = 2;
+        required uint64 size = 3;
+    }
+    required uint64 index = 1;
+    optional bytes value = 2;
+    repeated Node nodes = 3;
+    optional bytes signature = 4;
+}
+```
+
+#### Extension
+[msg-extension]: #msg-extension
+
+`type=15` Send a custom message.
+
+The `user-type` is an index into the `extensions` array sent in the Handshake. For instance, if two extensions were declared `['foo', 'bar']` then the `user-type` of `'foo'` would be `0` and of `'bar'` would be `1`.
+
+The `payload` may be any message content, as needed by the extension.
+
+```
+<varint user-type><payload>
+```
+
+
+## Run-Length Encoded Bitfield
+[run-length-encoded-bitfield]: #run-length-encoded-bitfield
+
+The Run-Length Encoded (RLE) bitfield is a series of compressed and uncompressed bit sequences.All sequences start with a header that is a varint. If the last bit is set in the varint (it is an odd number) then a header represents a compressed bit sequence. If the last bit is not set then a header represents an non compressed sequence.
+
+```
+compressed-sequence = varint(byte-length-of-sequence << 2 | bit << 1 | 1)
+uncompressed-sequence = varint(byte-length-of-bitfield << 1 | 0) + bitfield
+```
+
+
+## Block Tree Digest
+[block-tree-digest]: #block-tree-digest
+
+When asking for a block of data we want to reduce the amount of duplicate hashes that are sent back. To communicate which hashes we have, we just have to communicate two things: which uncles we have and whether or not we have any parent node that can verify the tree.
+
+Consider the following tree:
+
+```
+0
+  1
+2
+    3
+4
+  5
+6
+```
+
+If we want to fetch block 0, we need to communicate whether of not we already have the uncles (2, 5) and the parent (3). This information can be compressed into very small bit vector using the following scheme:
```

Comment:
I’m trying using figures to understand how this works. I still don’t grasp it, but here is where I’m at trying to draw the example shown in this section:

![hash_tree_digest](https://user-images.githubusercontent.com/34459172/51025955-4a0ac700-15f2-11e9-8344-da6806d5651b.png)

Reply:
My question here would be: If you just want to verify chunk 0, why not send this instead?

![hash_tree_digest](https://user-images.githubusercontent.com/34459172/51026701-fdc08680-15f3-11e9-8fbd-6bea78645bf4.png)

Reply:
I rewrote this section (and will push in a minute); please let me know if it clarified things. To answer a few questions directly:

The example of wanting to verify block 0 is bad; there's a direct signature of it. I believe that when you request a block, you always get the signature for the root hash as if that was the largest block index, and you're always trying to verify that signature, so I flipped the example to block 6.

This also answers the "how do you know how high the tree is" question.

You don't need hash 1 (in the old example) because you can calculate it from hash 0 and hash 2 (you calculate hash 2 directly from the data transmitted). I guess this is also why you don't need hash 3. Note that in other situations you would need other parent hashes.

I added a reference to the Hypercore DEP which is where the parent/uncle terminology should be defined (if it's unclear there we can take a patch to improve).

A more complex (non-symmetric) example would be nice, but also maybe more confusing here, where we're really just trying to explain the binary packing, not the whole Merkle hashing scheme?